src/programs/bammaskflags.1
.TH BAMMASKFLAGS 1 "July 2013" BIOBAMBAM
.SH NAME
bammaskflags - remove flags from alignments
.SH SYNOPSIS
.PP
.B bammaskflags
[options]
.SH DESCRIPTION
bammaskflags reads a BAM file from standard input, performs a bit wise and
with the flags field of each read/alignment and writes the resulting data to standard output in BAM format.
.PP
The following key=value pairs can be given:
.PP
.B level=<-1|0|1|9|11>:
set compression level of the output BAM file. Valid
values are
.IP -1:
zlib/gzip default compression level
.IP 0:
uncompressed
.IP 1:
zlib/gzip level 1 (fast) compression
.IP 9:
zlib/gzip level 9 (best) compression
.P
If libmaus has been compiled with support for igzip (see
https://software.intel.com/en-us/articles/igzip-a-high-performance-deflate-compressor-with-optimizations-for-genomic-data)
then an additional valid value is
.IP 11:
igzip compression
.PP
.B maskneg=<0>
mask of flags to be removed (a bit wise and is computed between the bit
complement of this number and the flag field of the reads/alignments in the
input BAM file). This value can be obtained by adding up the following
values of the flags:
.IP PAIRED (paired in sequencing):
1
.IP PROPER_PAIR (mapped as a proper pair):
2
.IP UNMAP (unmapped):
4
.IP MUNMAP (mate unmapped):
8
.IP REVERSE (mapped to the reverse strand):
16
.IP MREVERSE (mate mapped to the reverse strand):
32
.IP READ1 (first read of pair):
64
.IP READ2 (second read of pair):
128
.IP SECONDARY (secondary alignment):
256
.IP QCFAIL (failed quality control):
512
.IP DUP (duplicate):
1024
.IP SUPPLEMENTARY (supplementary):
2048
.PP
.B tmpfile=<filename>: 
prefix for temporary files. By default the temporary files are created in the current directory
.PP
.B md5=<0|1>:
md5 checksum creation for output file. Valid values are
.IP 0:
do not compute checksum. This is the default.
.IP 1:
compute checksum. If the md5filename key is set, then the checksum is
written to the given file. If md5filename is unset, then no checksum will be computed.
.PP
.B md5filename
file name for md5 checksum if md5=1.
.PP
.B index=<0|1>:
compute BAM index for output file. Valid values are
.IP 0:
do not compute BAM index. This is the default.
.IP 1:
compute BAM index. If the indexfilename key is set, then the BAM index is
written to the given file. If indexfilename is unset, then no BAM index will be computed.
.PP
.B indexfilename
file name for BAM index if index=1.
.PP
.B auxexists=[taglist]:
only mask records containing any of the given aux field tags. taglist is a
comma separated list of aux field names. Flag fields of records are only
filtered if the respective records contain any of the aux fields mentioned
in this list. By default all records are filtered. If e.g. auxexists=AS is
given, then only flag fields of records containing an AS aux field are masked.
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2013 German Tischler, \(co 2011-2013 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.87) unstable; urgency=medium

  * update to libmaus2 2.0.451

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 01 Mar 2018 13:15:37 +0100

biobambam2 (2.0.86) unstable; urgency=medium

  * add auxexists option in bammaskflags

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 28 Feb 2018 17:13:44 +0100

biobambam2 (2.0.85) unstable; urgency=medium

  * update to libmaus2 2.0.449

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 28 Feb 2018 15:41:16 +0100

biobambam2 (2.0.84) unstable; urgency=medium

  * update to libmaus2 2.0.448

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 23 Feb 2018 15:21:44 +0100

biobambam2 (2.0.83) unstable; urgency=medium

  * add queryname_HI sort mode in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 05 Jan 2018 12:08:05 +0100

biobambam2 (2.0.82) unstable; urgency=medium

  * use PTHREAD_MUTEX_RECURSIVE_NP or PTHREAD_MUTEX_RECURSIVE as available in bamadapterfind
  * add detection of PTHREAD_MUTEX_RECURSIVE_NP and PTHREAD_MUTEX_RECURSIVE in configure

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 14 Dec 2017 12:02:52 +0100

biobambam2 (2.0.81) unstable; urgency=medium

  * remove typename keywords inside non template class

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 05 Nov 2017 20:24:16 +0100

biobambam2 (2.0.80) unstable; urgency=medium

  * add minlength parameter in normalisefasta

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 24 Oct 2017 13:21:19 +0200

biobambam2 (2.0.79) unstable; urgency=medium

  * update to libmaus2 2.0.398

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 06 Oct 2017 11:37:27 +0200

biobambam2 (2.0.78) unstable; urgency=medium

  * update libmaus2 requirement

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 27 Sep 2017 10:56:56 +0200

biobambam2 (2.0.77) unstable; urgency=medium

  * add sorting by string aux field in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 27 Sep 2017 10:38:19 +0200

biobambam2 (2.0.76) unstable; urgency=medium

  * add rcsupport switch in bamsormadup

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 08 Aug 2017 16:49:12 +0200

biobambam2 (2.0.75) unstable; urgency=medium

  * add mc field in bammarkduplicatesopt

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 03 Aug 2017 12:24:54 +0200

biobambam2 (2.0.74) unstable; urgency=medium

  * add bamfillquery

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 02 Aug 2017 10:00:47 +0200

biobambam2 (2.0.73) unstable; urgency=medium

  * add bamfilterrefid
  * allow O key in bammerge (disable check for stdout being a terminal if O key is set)
  * add bamrefinterval program

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 20 Jun 2017 14:48:51 +0200

biobambam2 (2.0.72) unstable; urgency=medium

  * add output file prefix for default read group in bamtofastq

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Mar 2017 10:22:23 +0100

biobambam2 (2.0.71) unstable; urgency=medium

  * add outputperreadgrouprgsm and outputperreadgroupprefix options in bamtofastq

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Mar 2017 10:06:40 +0100

biobambam2 (2.0.70) unstable; urgency=medium

  * add bamscrapcount

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 07 Mar 2017 17:05:10 +0100

biobambam2 (2.0.69) unstable; urgency=medium

  * encode ref id of first read in file in bamexplode file names

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Jan 2017 13:32:37 +0100

biobambam2 (2.0.68) unstable; urgency=medium

  * added HD line header merging in bamcat

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Jan 2017 11:33:22 +0100

biobambam2 (2.0.67) unstable; urgency=medium

  * no adapter clipping if remaining sequence would be a single base (fix by Andrew Whitwham)
  * add bamauxmerge

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 26 Jan 2017 17:30:44 +0100

biobambam2 (2.0.66) unstable; urgency=medium

  * add parserbuffersize parameter in bamsormadup 
  * add optminpixeldif parameter passing in bamsormadup
  * add bamfilterlength program

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 05 Jan 2017 12:06:53 +0100

biobambam2 (2.0.65) unstable; urgency=medium

  * fix bug in bammarkduplicatesopt where last set of duplicates was no marked with od fields

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 30 Nov 2016 12:14:14 +0100

biobambam2 (2.0.64) unstable; urgency=medium

  * add addmatecigar option in bammarkduplicatesopt

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 29 Nov 2016 12:31:07 +0100

biobambam2 (2.0.63) unstable; urgency=medium

  * fix bug in bammarkduplicates/bammarkduplicates2 concerning input via pipe, markthreads>1 and rewritebam=2 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 28 Nov 2016 10:47:09 +0100

biobambam2 (2.0.62) unstable; urgency=medium

  * fix bin computation for unmapped reads in bamsormadup with SAM input 

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 25 Nov 2016 13:34:04 +0100

biobambam2 (2.0.61) unstable; urgency=medium

  * filter out broken alignments (cigar reaching past end of reference) in bamrecalculatecigar instead of quitting

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 21 Nov 2016 09:06:09 +0100

biobambam2 (2.0.60) unstable; urgency=medium

  * add bamnumericalindex program

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 17 Nov 2016 15:13:15 +0100

biobambam2 (2.0.59) unstable; urgency=medium

  * adapt bamparsort to new libmaus2 API
  * add numeric index construction in bamindex

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 14 Nov 2016 14:09:29 +0100

biobambam2 (2.0.58) unstable; urgency=medium

  * Update to libmaus2 version 2.0.268 (fixes for mate cigar string copying in bamsormadup)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 01 Nov 2016 10:35:57 +0100

biobambam2 (2.0.57) unstable; urgency=medium

  * Added an explicit iRODS disconnect to avoid the iRODS 4.19 crash. (patch by Andrew Whitwham)

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 18 Sep 2016 18:56:06 +0200

biobambam2 (2.0.56) unstable; urgency=medium


src/programs/bammaskflags.cpp
/**
    bambam
    Copyright (C) 2009-2013 German Tischler
    Copyright (C) 2011-2013 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include <libmaus2/bambam/CollatingBamDecoder.hpp>
#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>
#include <libmaus2/bambam/BamWriter.hpp>
#include <libmaus2/bambam/ProgramHeaderLineSet.hpp>
#include <libmaus2/util/ArgInfo.hpp>
#include <biobambam2/Licensing.hpp>

#include <config.h>

#include <libmaus2/lz/BgzfDeflateOutputCallbackMD5.hpp>
#include <libmaus2/bambam/BgzfDeflateOutputCallbackBamIndex.hpp>
static int getDefaultMD5() { return 0; }
static int getDefaultIndex() { return 0; }

static uint64_t getDefaultMaskNeg()
{
	return 0;
}

static int getDefaultLevel()
{
	return Z_DEFAULT_COMPRESSION;
}

int bammaskflags(::libmaus2::util::ArgInfo const & arginfo)
{
	uint64_t const maskpos = arginfo.getValue<uint64_t>("maskpos",0xFFFFUL);
	uint64_t const maskneg = arginfo.getValue<uint64_t>("maskneg",getDefaultMaskNeg());
	uint64_t const mask = maskpos & (~maskneg);

	::libmaus2::trie::LinearHashTrie<char,uint32_t>::shared_ptr_type Sauxexists;

	if ( arginfo.hasArg("auxexists") )
	{
		std::string const auxex = arginfo.getUnparsedValue("auxexists",std::string());
		std::deque<std::string> qauxex = ::libmaus2::util::stringFunctions::tokenize(auxex,std::string(","));
		std::vector<std::string> vauxex = std::vector<std::string>(qauxex.begin(),qauxex.end());


		for ( uint64_t i = 0; i < vauxex.size(); ++i )
			std::cerr << "auxexists " << vauxex[i] << std::endl;

		::libmaus2::trie::Trie<char> trienofailure;
		trienofailure.insertContainer(vauxex);
		::libmaus2::trie::LinearHashTrie<char,uint32_t>::unique_ptr_type LHTnofailure(trienofailure.toLinearHashTrie<uint32_t>());
		Sauxexists = ::libmaus2::trie::LinearHashTrie<char,uint32_t>::shared_ptr_type(LHTnofailure.release());
	}

	if ( mask )
	{
		std::cerr << "Keeping flags ";
		for ( uint64_t i = 1; i <= ::libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_FSUPPLEMENTARY; i <<= 1 )
			if ( mask & i )
				std::cerr << static_cast< ::libmaus2::bambam::BamFlagBase::bam_flags >(i) << ";";
		std::cerr << std::endl;
		std::cerr << "Erasing flags ";
		for ( uint64_t i = 1; i <= ::libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_FSUPPLEMENTARY; i <<= 1 )
			if ( !(mask & i) )
				std::cerr << static_cast< ::libmaus2::bambam::BamFlagBase::bam_flags >(i) << ";";
		std::cerr << std::endl;
	}
	else
	{
		std::cerr << "Erasing all flags." << std::endl;
	}

	int const level = libmaus2::bambam::BamBlockWriterBaseFactory::checkCompressionLevel(arginfo.getValue<int>("level",getDefaultLevel()));
	int const resetmatecoord = arginfo.getValue<int>("resetmatecoord",0);

	::libmaus2::bambam::BamDecoder BD(std::cin);
	::libmaus2::bambam::BamHeader const & bamheader = BD.getHeader();

	std::string const headertext(bamheader.text);

	// add PG line to header
	std::string const upheadtext = ::libmaus2::bambam::ProgramHeaderLineSet::addProgramLine(
		headertext,
		"bammaskflags", // ID
		"bammaskflags", // PN
		arginfo.commandline, // CL
		::libmaus2::bambam::ProgramHeaderLineSet(headertext).getLastIdInChain(), // PP
		std::string(PACKAGE_VERSION) // VN
	);
	// construct new header
	::libmaus2::bambam::BamHeader uphead(upheadtext);

	::libmaus2::bambam::BamAlignment & alignment = BD.getAlignment();

	/*
	 * start index/md5 callbacks
	 */
	std::string const tmpfilenamebase = arginfo.getValue<std::string>("tmpfile",arginfo.getDefaultTmpFileName());
	std::string const tmpfileindex = tmpfilenamebase + "_index";
	::libmaus2::util::TempFileRemovalContainer::addTempFile(tmpfileindex);

	std::string md5filename;
	std::string indexfilename;

	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > cbs;
	::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Pmd5cb;
	if ( arginfo.getValue<unsigned int>("md5",getDefaultMD5()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo) != std::string() )
			md5filename = libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo);
		else
			std::cerr << "[V] no filename for md5 given, not creating hash" << std::endl;

		if ( md5filename.size() )
		{
			::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Tmd5cb(new ::libmaus2::lz::BgzfDeflateOutputCallbackMD5);
			Pmd5cb = UNIQUE_PTR_MOVE(Tmd5cb);
			cbs.push_back(Pmd5cb.get());
		}
	}
	libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Pindex;
	if ( arginfo.getValue<unsigned int>("index",getDefaultIndex()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo) != std::string() )
			indexfilename = libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo);
		else
			std::cerr << "[V] no filename for index given, not creating index" << std::endl;

		if ( indexfilename.size() )
		{
			libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Tindex(new libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex(tmpfileindex));
			Pindex = UNIQUE_PTR_MOVE(Tindex);
			cbs.push_back(Pindex.get());
		}
	}
	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > * Pcbs = 0;
	if ( cbs.size() )
		Pcbs = &cbs;
	/*
	 * end md5/index callbacks
	 */

	::libmaus2::bambam::BamWriter::unique_ptr_type writer(new ::libmaus2::bambam::BamWriter(std::cout,uphead,level,Pcbs));
	libmaus2::autoarray::AutoArray < std::pair<uint8_t,uint8_t> > Aaux;

	while ( BD.readAlignment() )
	{
		bool dofilter = false;

		if ( !Sauxexists )
		{
			dofilter = true;
		}
		else
		{
			uint64_t const numaux = alignment.enumerateAuxTags(Aaux);

			for ( uint64_t i = 0; i < numaux; ++i )
			{
				uint8_t const A[3] = { Aaux[i].first, Aaux[i].second, 0 };
				dofilter = dofilter || ( Sauxexists->searchCompleteNoFailureZ(&A[0]) != -1 );
			}
		}

		if ( dofilter )
			alignment.putFlags(alignment.getFlags() & mask);

		if ( resetmatecoord )
		{
			alignment.putNextRefId(-1);
			alignment.putNextPos(-1);
		}
		alignment.serialise(writer->getStream());
	}

	writer.reset();

	if ( Pmd5cb )
	{
		Pmd5cb->saveDigestAsFile(md5filename);
	}
	if ( Pindex )
	{
		Pindex->flush(std::string(indexfilename));
	}

	return EXIT_SUCCESS;
}

int main(int argc, char * argv[])
{
	try
	{
		::libmaus2::util::ArgInfo const arginfo(argc,argv);

		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if (
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if (
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				std::cerr << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;

				std::vector< std::pair<std::string,std::string> > V;

				V.push_back ( std::pair<std::string,std::string> ( "level=<["+::biobambam2::Licensing::formatNumber(getDefaultLevel())+"]>", libmaus2::bambam::BamBlockWriterBaseFactory::getBamOutputLevelHelpText() ) );
				V.push_back ( std::pair<std::string,std::string> ( "maskneg=<["+::biobambam2::Licensing::formatNumber(getDefaultMaskNeg())+"]>", "flag masking bitmask (bits set are removed)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "md5=<["+::biobambam2::Licensing::formatNumber(getDefaultMD5())+"]>", "create md5 check sum (default: 0)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "md5filename=<filename>", "file name for md5 check sum (default: extend output file name)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "index=<["+::biobambam2::Licensing::formatNumber(getDefaultIndex())+"]>", "create BAM index (default: 0)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "indexfilename=<filename>", "file name for BAM index file (default: extend output file name)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "tmpfile=<filename>", "prefix for temporary files, default: create files in current directory" ) );
				V.push_back ( std::pair<std::string,std::string> ( "auxexists=[]", "only mask records containing any of the given aux field tags" ) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;
				return EXIT_SUCCESS;
			}

		return bammaskflags(arginfo);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}
